File: clj/src/h3m_lwp_clj/assets.clj
(ns h3m-lwp-clj.assets
  (:import
   [com.badlogic.gdx Gdx]
   [com.badlogic.gdx.graphics Texture]
   [com.badlogic.gdx.graphics.g2d TextureAtlas]
   [com.badlogic.gdx.assets AssetManager]
   [com.badlogic.gdx.assets.loaders TextureAtlasLoader$TextureAtlasParameter]
   [com.badlogic.gdx.utils Array])
  (:require
   [h3m-lwp-clj.consts :as consts]))


(def asset-manager (atom nil))


(defn file-exists?
  [file-path]
  (->> file-path
       (.local Gdx/files)
       (.exists)))


(defn assets-ready?
  []
  (and
   (file-exists? consts/atlas-file-name)
   (file-exists? consts/png-file-name)))


(defn load-atlas
  [^AssetManager asset-manager]
  (doto asset-manager
    (.load
     consts/atlas-file-name
     TextureAtlas
     (new TextureAtlasLoader$TextureAtlasParameter true))
    (.finishLoadingAsset consts/atlas-file-name)
    (.finishLoading)))


(defn validate-atlas
  [^AssetManager asset-manager]
  (let [^TextureAtlas atlas (.get asset-manager consts/atlas-file-name)
        regions (.getRegions atlas)
        assets-count (.-size regions)
        ; TODO check by defs used in map
        enough-assets? (>= assets-count consts/minimal-lod-objects-count)]
    (if (not enough-assets?)
      (do
        (.dispose asset-manager)
        nil)
      asset-manager)))


(defn try-load-atlas
  []
  (when (assets-ready?)
    (let [manager (some->
                   (new AssetManager)
                   (load-atlas)
                   (validate-atlas))]
      (when manager
        (Texture/setAssetManager manager)
        (reset! asset-manager manager)
        true))))


(defn get-def-frames
  ^Array
  [def-name]
  (let [^TextureAtlas atlas (.get ^AssetManager @asset-manager consts/atlas-file-name)
        ^Array frames (.findRegions atlas def-name)]
    (when (.isEmpty frames)
      (println (format "def %s not found in atlas" def-name)))
    frames))


(defn get-terrain-sprite
  ^Array
  [def-name index]
  (let [frames (get-def-frames def-name)]
    (if (.isEmpty frames)
      frames
      (doto (new Array 1)
        (.add (.get frames index))))))

File: core/src/main/kotlin/core/Assets.kt
package com.homm3.livewallpaper.core

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.assets.loaders.TextureAtlasLoader
import com.badlogic.gdx.assets.loaders.resolvers.InternalFileHandleResolver
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.I18NBundle
import com.homm3.livewallpaper.parser.formats.H3m
import ktx.assets.load
import ktx.collections.gdxArrayOf
import java.util.*

class Assets {
    val manager = AssetManager(InternalFileHandleResolver()).also {
        it.setLoader(H3m::class.java, H3mLoader(it.fileHandleResolver))
    }
    private val internalManager = AssetManager(InternalFileHandleResolver())
    lateinit var skin: Skin
    lateinit var i18n: I18NBundle

    init {
        Texture.setAssetManager(manager)
    }

    fun loadUiAssets() {
        skin = internalManager.load<Skin>(Constants.Assets.SKIN_PATH).let {
            it.finishLoading()
            it.asset
        }
        i18n = internalManager.load<I18NBundle>(Constants.Assets.I18N_PATH).let {
            I18NBundle.setExceptionOnMissingKey(false)
            it.finishLoading()
            it.asset
        }
    }

    fun isGameAssetsLoaded(): Boolean {
        return manager.isLoaded(Constants.Assets.ATLAS_PATH)
    }

    fun isGameAssetsAvailable(): Boolean {
        val isAssetsReady = Gdx.app
            .getPreferences(Constants.Preferences.PREFERENCES_NAME)
            .getBoolean(Constants.Preferences.IS_ASSETS_READY_KEY)
        val filesExists = Gdx.files.local(Constants.Assets.ATLAS_PATH).exists()

        return isAssetsReady && filesExists
    }

    fun loadGameAssets() {
        manager
            .load<TextureAtlas>(
                Constants.Assets.ATLAS_PATH,
                TextureAtlasLoader.TextureAtlasParameter(true)
            )

        manager.finishLoading()
    }

    fun getTerrainFrames(defName: String, index: Int): Array<TextureAtlas.AtlasRegion> {
        manager
            .get<TextureAtlas>(Constants.Assets.ATLAS_PATH)
            .findRegions("$defName/$index")
            .run {
                return if (isEmpty) {
                    println("Can't find terrain def $defName/$index")
                    return gdxArrayOf(Constants.Assets.emptyPixmap)
                } else {
                    this
                }
            }
    }

    fun getObjectFrames(defName: String): Array<TextureAtlas.AtlasRegion> {
        val name = defName.toLowerCase(Locale.ROOT).removeSuffix(".def");

        manager
            .get<TextureAtlas>(Constants.Assets.ATLAS_PATH)
            .findRegions(name)
            .run {
                return if (isEmpty) {
                    println("Can't find objects def $name")
                    return gdxArrayOf(Constants.Assets.emptyPixmap)
                } else {
                    this
                }
            }
    }
}